<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Testing-pandas-reading-from-&quot;data&quot;-directory" data-toc-modified-id="Testing-pandas-reading-from-&quot;data&quot;-directory-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Testing pandas reading from "data" directory</a></span></li><li><span><a href="#Testing-connection-to-our-PostgreSQL-database" data-toc-modified-id="Testing-connection-to-our-PostgreSQL-database-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Testing connection to our PostgreSQL database</a></span></li><li><span><a href="#Lets-test-getting-data-from-web-page" data-toc-modified-id="Lets-test-getting-data-from-web-page-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Lets test getting data from web page</a></span></li></ul></div>

## Testing pandas reading from "data" directory

In [1]:
from IPython.display import display, HTML
display(HTML("<style>.container { width:95% !important; }</style>"))

In [2]:
import os
import pandas as pd

df = pd.read_csv("sibin_dev/data/raw/argos_categories_lvl1.csv", delimiter="|")
print(df)


          category_name                                       category_url
0      Home & furniture  https://www.argos.co.uk/browse/home-and-furnit...
1            Technology  https://www.argos.co.uk/browse/technology/c:29...
2            Appliances  https://www.argos.co.uk/browse/appliances/c:78...
3                  Toys       https://www.argos.co.uk/browse/toys/c:30299/
4          Garden & DIY  https://www.argos.co.uk/browse/garden-and-diy/...
5   Jewellery & watches  https://www.argos.co.uk/browse/jewellery-and-w...
6      Sports & Leisure  https://www.argos.co.uk/browse/sports-and-leis...
7       Health & Beauty  https://www.argos.co.uk/browse/health-and-beau...
8        Baby & Nursery  https://www.argos.co.uk/browse/baby-and-nurser...
9              Clothing   https://www.argos.co.uk/browse/clothing/c:29131/
10      Gifts & Parties      https://www.argos.co.uk/browse/gifts/c:29164/


## Testing connection to our PostgreSQL database


In [ ]:
import psycopg2

# Connect to the PostgreSQL database
conn = psycopg2.connect(
    host="localhost",
    port="5432",
    database="sibin",
    user="sibinuser",
    password="sib123"
)

# Create a cursor object to execute SQL queries
cur = conn.cursor()

# Execute a sample SQL query
cur.execute("SELECT * FROM users")

# Fetch the query result
result = cur.fetchall()

# Print the query result
print(result)

# Close the cursor and connection objects
cur.close()
conn.close()


## Lets test getting data from web page 

I know we mentioned that we will scrappe sibenik.in but we won't have much data if we do that. This is just first page but ideas is to get all houses for sale in sibenik from https://www.index.hr/oglasi/prodaja-kuca 




In [30]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# This is just copy pasted from Chrome after setting Šibenik as county and houses. Copy paste this link into Chrome to see how it looks. 
url = 'https://www.index.hr/oglasi/prodaja-kuca/gid/3276?pojam=&sortby=1&elementsNum=10&cijenaod=0&cijenado=15000000&tipoglasa=1&btnTipOglasa=1&pojamZup=1165&naselje=&attr_Int_485=&attr_Int_899=&attr_bit_stan=&attr_bit_brojEtaza=&attr_Int_529=&attr_Int_907=&attr_Int_486=&attr_Int_903=&attr_Int_978=&attr_Int_1334=&attr_Int_979_od=&attr_Int_979_do=&attr_bit_eneregetskiCertifikat=&vezani_na=485-899_529-907_486-903_978-1334'
response = requests.get(url)

# Render HTML into object in which we can search and extract by HTML tags"
soup = BeautifulSoup(response.content, 'html.parser')

# Find our ads
ads = soup.find_all('div', class_='OglasiRezHolder')

# Define list where we will store all ads data and latter we'll build pandas dataframe from this
data = []
for i, ad in enumerate(ads, 1):
    print("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx START AD PROCESSING xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n")
    print(f"Procesing {i}, ad HTML content looks like this, lets get data after this:\n", ad)
    classes = ad.get('class', [])
    # Checking if "OglasiRezHolder" is only class in <div class="OglasiRezHolder"> if it's not than this is probably advertisement (Google add or something like that)
    if len(classes) == 1 and 'OglasiRezHolder' in classes:
        print(f'\nProcessing ad {i}/{len(ads)}...')
        title = ad.find('span', class_=lambda c: c.startswith('title')).text.strip()
        print(f'Title: {title}')
        img_tag = ad.find('img', class_='adexFoundElement-0')
        img_url = img_tag['src'] if img_tag else ''
        print(f'Image URL: {img_url}')
        description = ad.find('span', class_='lead').text.strip()
        print(f'Description: {description}')
        m2AroundHouse = ad.find('ul', class_='tags').text.strip().split(':')[1].strip()
        print(f'm2AroundHouse: {m2AroundHouse}')
        county = ad.find('li', class_='icon-marker').text.strip()
        print(f'County: {county}')
        published_ago = ad.find('li', class_='icon-time').text.strip()
        print(f'Published ago: {published_ago}')
        price_EUR = ad.find('span', class_='price').find('span').text.strip()
        print(f'Price EUR: {price_EUR}')
        price_kn = ad.find('span', class_='price').text.strip().split('~')[1].strip()
        print(f'Price kn: {price_kn}')
        ad_url = ad.find('a', class_='result')['href']
        print(f'Ad URL: {ad_url}\n')

        data.append({
            'Title': title,
            'Image': img_url,
            'Description': description,
            'm2AroundHouse': m2AroundHouse,
            'County': county,
            'Published_ago': published_ago,
            'Price_EUR': price_EUR,
            'Price_kn': price_kn,
            'Ad_URL': ad_url
        })
    else:
        print(f'\nSkipping ad {i}/{len(ads)}...')
        # If word "Banner" is in name of HTML class 
        if any('Banner' in c for c in classes):
            print('This is an advertisement and not a house ad listing')
        else:
            other_classes = [c for c in classes if c != 'OglasiRezHolder']
            print(f'Classes: {other_classes}')

df = pd.DataFrame(data)
display(df)


xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx START AD PROCESSING xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Procesing 1, ad HTML content looks like this, lets get data after this:
 <div class="OglasiRezHolder">
<a class="result" href="https://www.index.hr/oglasi/kuca-s-bazenom-212-32-m2--vodice/oid/3925344">
<figure class="result_photo" data-x="2">
<img alt="Kuća s bazenom 212.32 m2, Vodice" src="https://www.index.hr/oglasi/userdocsimages/oglas/_2023//3//17///3925344/k21945_20230317012934534_0_.jpg?preset=oglas-slika-view"/>
</figure>
<span class="result_body">
<span class="head">
<span class="title px18">
Kuća s bazenom 212.32 m2, Vodice
<span class="addFavorit icon-heart tooltip_bottom" title="Spremi oglas">
</span>
</span>
<span class="lead" data-fff="bb">Kuća s bazenom 212.32 m2, VodiceKuća s bazenom 212,32 m2, VodiceNa mirnoj lokaciji u blizini samog centra Vodica, turist</span>
<ul class="tags hide-on-small-only"><li>
m2 okućnice
 : 295.00
</li>
</ul>
</span>
<span class="foot">
<ul cl

Title Image  \
0                   Kuća s bazenom 212.32 m2, Vodice         
1                   Kuća sa bazenom 197.77m2, Vodice         
2            Obiteljska kuća s okućnicom, 200m2 Knin         
3            Vila s bazenom, 223m2, okolica Šibenika         
4            Rogoznica 200 m od mora villa s bazenom         
5           Rogoznica 100 m od mora apartmanska kuća         
6       Vodice, Vlila sa bazenom i velikom okućnicom         
7  DALMACIJA, PRIMOŠTEN - Prekrasna vila sa čarob...         
8                    Rogoznica - kuća sa 2 apartmana         
9  Vila iz snova u Dalmaciji s bazenom i pogledom...         

                                         Description  \
0  Kuća s bazenom 212.32 m2, VodiceKuća s bazenom...   
1  Kuća sa bazenom 197.77m2, VodiceKuća s bazenom...   
2  Obiteljska kuća s okućnicom, 200m2 KninKuća Kn...   
3  Vila s bazenom, 223m2, okolica ŠibenikaVila s ...   
4  U Rogoznici malom turističkom mjestu udaljenom...   
5  Rogoznica 100 m od mora prekrasna apartmanska ...   
6  U mirnom okruženju, idealnom za uživanje u mir...   
7  DALMACIJA, PRIMOŠTEN - Prekrasna vila sa čarob...   
8  Posredujemo u prodaji samostojeće kuće na moru...   
9  cca 200 kvadrata stambenog prostora3 spavaće s...   

                m2AroundHouse            County          Published_ago  \
0                      295.00  Šibensko-kninska  Objavljeno prije 24 h   
1                      295.00  Šibensko-kninska  Objavljeno prije 24 h   
2                      390.00  Šibensko-kninska  Objavljeno prije 24 h   
3                     1000.00  Šibensko-kninska  Objavljeno prije 24 h   
4   9.00\r\n\n\r\nm2 okućnice  Šibensko-kninska  Objavljeno prije 24 h   
5  15.00\r\n\n\r\nm2 okućnice  Šibensko-kninska  Objavljeno prije 24 h   
6   6.00\r\n\n\r\nm2 okućnice  Šibensko-kninska  Objavljeno prije 24 h   
7   6.00\r\n\n\r\nm2 okućnice  Šibensko-kninska  Objavljeno prije 24 h   
8            2.00\r\n\n\r\nm2  Šibensko-kninska  Objavljeno prije 24 h   
9   4.00\r\n\n\r\nm2 okućnice  Šibensko-kninska    Objava  16.03.2023.   

     Price_EUR       Price_kn  \
0    400.000 €   3.013.800 kn   
1    390.000 €   2.938.455 kn   
2    156.000 €   1.175.382 kn   
3    550.000 €   4.143.975 kn   
4    625.000 €   4.709.063 kn   
5    880.000 €   6.630.360 kn   
6  1.400.000 €  10.548.300 kn   
7    780.000 €   5.876.910 kn   
8    355.000 €   2.674.748 kn   
9    690.000 €   5.198.805 kn   

                                              Ad_URL  
0  https://www.index.hr/oglasi/kuca-s-bazenom-212...  
1  https://www.index.hr/oglasi/kuca-sa-bazenom-19...  
2  https://www.index.hr/oglasi/obiteljska-kuca-s-...  
3  https://www.index.hr/oglasi/vila-s-bazenom--22...  
4  https://www.index.hr/oglasi/rogoznica-200-m-od...  
5  https://www.index.hr/oglasi/rogoznica-100-m-od...  
6  https://www.index.hr/oglasi/vodice--vlila-sa-b...  
7  https://www.index.hr/oglasi/dalmacija--primost...  
8  https://www.index.hr/oglasi/rogoznica---kuca-s...  
9  https://www.index.hr/oglasi/vila-iz-snova-u-da...

##                                    My own work

In [12]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

url = 'https://www.index.hr/oglasi/motorni-brodovi/gid/3573'
response = requests.get(url)
soup = BeautifulSoup(response.content, 'html.parser')

# extract data from the HTML
data = []
for ad in soup.find_all('div', {'class': 'search-ads__item'}):
    title = ad.find('h2', {'class': 'search-ads__title'}).text.strip()
    price = ad.find('div', {'class': 'search-ads__price'}).text.strip()
    data.append([title, price])

# create a pandas dataframe
df = pd.DataFrame(data, columns=['Title', 'Price'])
print(df)

Empty DataFrame
Columns: [Title, Price]
Index: []


In [24]:

import requests as r
from bs4 import BeautifulSoup as soup
url = 'https://www.index.hr/oglasi/motorni-brodovi/gid/3573'
data = r.get(url)
data
page_soup = soup(data.text, 'html.parser')
print(page_soup.prettify())
data2 = []
ad = page_soup.find('div', class_='OglasiRezHolder')
title_tag = ad.find('span', class_=lambda c: c.startswith('title')).text.strip()
if title_tag:
    title = title_tag.text.strip()
else:
    title = 'Not stated'
print(f'Title: {title}')
description = ad.find('span', class_='lead').text.strip()
print(f'Description: {description}')
price_EUR = ad.find('span', class_='price').find('span').text.strip()
print(f'Price EUR: {price_EUR}')
price_kn = ad.find('span', class_='price').text.strip().split('~')[1].strip()
print(f'Price kn: {price_kn}')
county_tag = ad.find('li', class_='icon-marker')
if county_tag:
    county = county_tag.text.strip()
else:
    county = 'Not stated'
print(f'County: {county}')
ad_url = ad.find('a', class_='result')['href']
print(f'Ad URL: {ad_url}\n')

data2.append({
            'Title': title_tag,
            'Description': description,
            'Price_EUR': price_EUR,
            'Price_kn': price_kn,
            'County': county_tag,        
            'Ad_URL': ad_url
        })

import pandas as pd
df = pd.DataFrame(data2)
display(df)

<!DOCTYPE html>
<html>
 <head>
  <meta charset="utf-8"/>
  <meta content="1697489853797759" property="fb:app_id"/>
  <title>
   Motorni brodovi | Plovila na prodaju
  </title>
  <meta content="Motorni brodovi | Plovila na prodaju" property="og:title"/>
  <meta content="https://www.index.hr/oglasi/motorni-brodovi/gid/3573" property="og:url"/>
  <meta content="Pretražite oglase, saznajte cijenu i pogledajte slike. – Index Oglasi." name="description"/>
  <meta content="www.index.hr/oglasi" property="og:site_name"/>
  <meta content="https://www.index.hr/oglasi/img/IndexOglasiLogo.jpg" property="og:image"/>
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=3.0, minimum-scale=1.0" name="viewport"/>
  <meta content="telephone=no" name="format-detection">
   <link href="https://www.index.hr/oglasi/favicon.ico" rel="SHORTCUT ICON" type="image/ico"/>
   <link href="https://www.index.hr/oglasi/favicon.png" rel="icon" type="image/png"/>
   <style>
    @charset "UTF-8";.selectat

Title                Description Price_EUR    Price_kn  \
0  Quicksilver commander 635  Quicksilver commander 635  17.000 €  128.087 kn   

  County                                             Ad_URL  
0   None  https://www.index.hr/oglasi/quicksilver-comman...

In [34]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd

url = 'https://www.index.hr/oglasi/motorni-brodovi/gid/3573'
data2 = []

while True:
    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title'))
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        
        price_EUR = ad.find('span', class_='price').find('span').text.strip()
        print(f'Price EUR: {price_EUR}')
        price_kn = ad.find('span', class_='price').text.strip().split('~')[1].strip()
        print(f'Price kn: {price_kn}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_url = ad.find('a', class_='result')['href']
        print(f'Ad URL: {ad_url}\n')

        data2.append({
            'Title': title,
            'Price_EUR': price_EUR,
            'Price_kn': price_kn,
            'County': county,
            'Ad_URL': ad_url
        })

    # Check if there is a next page
    next_page = page_soup.find('a', {'title': 'Sljedeća stranica'})
    if not next_page:
        break
    
        url = next_page['href']

df = pd.DataFrame(data2)
display(df)

Title: Quicksilver commander 635
Price EUR: 17.000 €
Price kn: 128.087 kn
County: Not stated
Ad URL: https://www.index.hr/oglasi/quicksilver-commander-635/oid/3932897

Title: Not stated


AttributeError: 'NoneType' object has no attribute 'find'

In [35]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd

url = 'https://www.index.hr/oglasi/motorni-brodovi/gid/3573'
data2 = []

while True:
    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title')).text.strip()
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        description = ad.find('span', class_='lead').text.strip()
        print(f'Description: {description}')
        price_EUR_tag = ad.find('span', class_='price')
        if price_EUR_tag:
            price_EUR = price_EUR_tag.find('span').text.strip()
        else:
            price_EUR = 'Not stated'
        print(f'Price EUR: {price_EUR}')
        price_kn_tag = ad.find('span', class_='price')
        if price_kn_tag:
            price_kn = price_kn_tag.text.strip().split('~')[1].strip()
        else:
            price_kn = 'Not stated'
        print(f'Price kn: {price_kn}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_url = ad.find('a', class_='result')['href']
        print(f'Ad URL: {ad_url}\n')

        data2.append({
            'Title': title,
            'Description': description,
            'Price_EUR': price_EUR,
            'Price_kn': price_kn,
            'County': county,
            'Ad_URL': ad_url
        })

    # Check if there is a next page
    next_page = page_soup.find('a', {'title': 'Sljedeća stranica'})
    if not next_page:
        break
    
    url = next_page['href']

df = pd.DataFrame(data2)
display(df)

AttributeError: 'str' object has no attribute 'text'

In [38]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd

url = 'https://www.index.hr/oglasi/motorni-brodovi/gid/3573'
data2 = []

while True:
    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title')).text.strip()
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        description = ad.find('span', class_='lead').text.strip()
        print(f'Description: {description}')
        price_EUR_tag = ad.find('span', class_='price').find('span')
        if price_EUR_tag:
            price_EUR = price_EUR_tag.text.strip()
        else:
            price_EUR = 'Not stated'
        print(f'Price EUR: {price_EUR}')
        price_kn_tag = ad.find('span', class_='price')
        if price_kn_tag:
            price_kn = price_kn_tag.text.strip().split('~')[1].strip()
        else:
            price_kn = 'Not stated'
        print(f'Price kn: {price_kn}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_url = ad.find('a', class_='result')['href']
        print(f'Ad URL: {ad_url}\n')
        
        data2.append({
            'Title': title,
            'Description': description,
            'Price_EUR': price_EUR,
            'Price_kn': price_kn,
            'County': county,
            'Ad_URL': ad_url
        })

    # Check if there is a next page
    next_page = page_soup.find('a', {'title': 'Sljedeća stranica'})
    if not next_page:
        break
    
    url = next_page['href']

df = pd.DataFrame(data2)
display(df)

AttributeError: 'str' object has no attribute 'text'

In [9]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd
for page number in range (1, 89):
url = 'https://www.index.hr/oglasi/motorni-brodovi/gid/3573'
data2 = []

while True:
    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title'))
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        price_EUR_tag = ad.find('span', class_='price')
        if price_EUR_tag:
            price_EUR = price_EUR_tag.find('span').text.strip()
        else:
            price_EUR = 'Not stated'
        print(f'Price EUR: {price_EUR}')
        price_kn_tag = ad.find('span', class_='price')
        if price_kn_tag:
            price_kn = price_kn_tag.find('span').text.strip()
        else:
            price_kn = 'Not stated'
        print(f'Price kn: {price_kn}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_link = ad.find('a', class_='result')
        if ad_link:
            ad_url = ad_link['href']
        else:
            ad_url = 'Not stated'
        print(f'Ad URL: {ad_url}\n')

        data2.append({
        'Title': title,
        'Price_EUR': price_EUR,
        'Price_kn': price_kn,
        'County': county,
        'Ad_URL': ad_url
    })
    next_page = page_soup.find('a', {'title': 'Sljedeća stranica'})
    if not next_page:
        break
    
    url = next_page['href']

df = pd.DataFrame(data2)
display(df)

SyntaxError: invalid syntax (1978798355.py, line 4)

In [13]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd
data2 = []
for page_number in range (1, 89):
        url = f'https://www.index.hr/oglasi/motorni-brodovi/gid/3573?page={page_number}'


        data = r.get(url)
        page_soup = soup(data.text, 'html.parser')
        ads = page_soup.find_all('div', class_='OglasiRezHolder')
        for ad in ads:
            title_tag = ad.find('span', class_=lambda c: c.startswith('title'))
            if title_tag:
                title = title_tag.text.strip()
            else:
                title = 'Not stated'
            print(f'Title: {title}')
            price_EUR_tag = ad.find('span', class_='price')
            if price_EUR_tag:
                price_EUR = price_EUR_tag.find('span').text.strip()
            else:
                price_EUR = 'Not stated'
            print(f'Price EUR: {price_EUR}')
            price_kn_tag = ad.find('span', class_='price')
            if price_kn_tag:
                price_kn = price_kn_tag.find('span').text.strip()
            else:
                price_kn = 'Not stated'
            print(f'Price kn: {price_kn}')
            county_tag = ad.find('li', class_='icon-marker')
            if county_tag:
                county = county_tag.text.strip()
            else:
                county = 'Not stated'
            print(f'County: {county}')
            ad_link = ad.find('a', class_='result')
            if ad_link:
                ad_url = ad_link['href']
            else:
                ad_url = 'Not stated'
            print(f'Ad URL: {ad_url}\n')

            data2.append({
            'Title': title,
            'Price_EUR': price_EUR,
            'Price_kn': price_kn,
            'County': county,
            'Ad_URL': ad_url
        })
        next_page = page_soup.find('a', {'title': 'Sljedeća stranica'})
        if not next_page:
            break
    
        url = next_page['href']

df = pd.DataFrame(data2)
display(df.tail())

Title: Mariah SC 19
Price EUR: 22.000 €
Price kn: 22.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/mariah-sc-19/oid/3934322

Title: Not stated
Price EUR: Not stated
Price kn: Not stated
County: Not stated
Ad URL: Not stated

Title: Jeanneau Merry Fisher 795
Price EUR: 77.000 €
Price kn: 77.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/jeanneau-merry-fisher-795/oid/3934289

Title: Quicksilver commander 635
Price EUR: 17.000 €
Price kn: 17.000 €
County: Not stated
Ad URL: https://www.index.hr/oglasi/quicksilver-commander-635/oid/3932897

Title: Idea Marine 58 Open b+Yamaha 115 KS - ekskl. uvoznik i distributer
Price EUR: 36.035 €
Price kn: 36.035 €
County: Grad Zagreb
Ad URL: https://www.index.hr/oglasi/idea-marine-58-open-b-yamaha-115-ks-ekskl-uvoznik-i-distributer/oid/3932648

Title: Gliser norveška gradnja + Honda BF 150 4t
Price EUR: 13.000 €
Price kn: 13.000 €
County: Šibensko-kninska
Ad URL: https://www.index.hr/oglasi/gliser-norveska-g

Title   Price_EUR    Price_kn  \
8                     jeanneau leader 705    23.000 €    23.000 €   
9               Bellingardo 630 Sea Ghost     6.500 €     6.500 €   
10                     Gobbi pilotina 600    11.000 €    11.000 €   
11  Axopar 28 Cabin! Samo 75 radnih sati!   175.000 €   175.000 €   
12                             Not stated  Not stated  Not stated   

                County                                             Ad_URL  
8   Primorsko-goranska  https://www.index.hr/oglasi/jeanneau-leader-70...  
9   Primorsko-goranska  https://www.index.hr/oglasi/bellingardo-630-se...  
10  Primorsko-goranska  https://www.index.hr/oglasi/gobbi-pilotina-600...  
11  Primorsko-goranska  https://www.index.hr/oglasi/axopar-28-cabin-sa...  
12          Not stated                                         Not stated

In [14]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd

data2 = []
for page_number in range(1, 89):
    url = f'https://www.index.hr/oglasi/motorni-brodovi/gid/3573?page={page_number}'

    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title'))
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        price_EUR_tag = ad.find('span', class_='price')
        if price_EUR_tag:
            price_EUR = price_EUR_tag.find('span').text.strip()
        else:
            price_EUR = 'Not stated'
        print(f'Price EUR: {price_EUR}')
        price_kn_tag = ad.find('span', class_='price')
        if price_kn_tag:
            price_kn = price_kn_tag.find('span').text.strip()
        else:
            price_kn = 'Not stated'
        print(f'Price kn: {price_kn}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_link = ad.find('a', class_='result')
        if ad_link:
            ad_url = ad_link['href']
        else:
            ad_url = 'Not stated'
        print(f'Ad URL: {ad_url}\n')

        data2.append({
            'Title': title,
            'Price_EUR': price_EUR,
            'Price_kn': price_kn,
            'County': county,
            'Ad_URL': ad_url
        })

df = pd.DataFrame(data2)
display(df)

Title: Mariah SC 19
Price EUR: 22.000 €
Price kn: 22.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/mariah-sc-19/oid/3934322

Title: Not stated
Price EUR: Not stated
Price kn: Not stated
County: Not stated
Ad URL: Not stated

Title: Jeanneau Merry Fisher 795
Price EUR: 77.000 €
Price kn: 77.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/jeanneau-merry-fisher-795/oid/3934289

Title: Quicksilver commander 635
Price EUR: 17.000 €
Price kn: 17.000 €
County: Not stated
Ad URL: https://www.index.hr/oglasi/quicksilver-commander-635/oid/3932897

Title: Idea Marine 58 Open b+Yamaha 115 KS - ekskl. uvoznik i distributer
Price EUR: 36.035 €
Price kn: 36.035 €
County: Grad Zagreb
Ad URL: https://www.index.hr/oglasi/idea-marine-58-open-b-yamaha-115-ks-ekskl-uvoznik-i-distributer/oid/3932648

Title: Gliser norveška gradnja + Honda BF 150 4t
Price EUR: 13.000 €
Price kn: 13.000 €
County: Šibensko-kninska
Ad URL: https://www.index.hr/oglasi/gliser-norveska-g

Title   Price_EUR  \
0                                          Mariah SC 19    22.000 €   
1                                            Not stated  Not stated   
2                             Jeanneau Merry Fisher 795    77.000 €   
3                             Quicksilver commander 635    17.000 €   
4     Idea Marine 58 Open b+Yamaha 115 KS - ekskl. u...    36.035 €   
...                                                 ...         ...   
1139                                jeanneau leader 705    23.000 €   
1140                          Bellingardo 630 Sea Ghost     6.500 €   
1141                                 Gobbi pilotina 600    11.000 €   
1142              Axopar 28 Cabin! Samo 75 radnih sati!   175.000 €   
1143                                         Not stated  Not stated   

        Price_kn              County  \
0       22.000 €  Primorsko-goranska   
1     Not stated          Not stated   
2       77.000 €  Primorsko-goranska   
3       17.000 €          Not stated   
4       36.035 €         Grad Zagreb   
...          ...                 ...   
1139    23.000 €  Primorsko-goranska   
1140     6.500 €  Primorsko-goranska   
1141    11.000 €  Primorsko-goranska   
1142   175.000 €  Primorsko-goranska   
1143  Not stated          Not stated   

                                                 Ad_URL  
0     https://www.index.hr/oglasi/mariah-sc-19/oid/3...  
1                                            Not stated  
2     https://www.index.hr/oglasi/jeanneau-merry-fis...  
3     https://www.index.hr/oglasi/quicksilver-comman...  
4     https://www.index.hr/oglasi/idea-marine-58-ope...  
...                                                 ...  
1139  https://www.index.hr/oglasi/jeanneau-leader-70...  
1140  https://www.index.hr/oglasi/bellingardo-630-se...  
1141  https://www.index.hr/oglasi/gobbi-pilotina-600...  
1142  https://www.index.hr/oglasi/axopar-28-cabin-sa...  
1143                                         Not stated  

[1144 rows x 5 columns]

In [15]:
df.to_csv('boat_ads.csv', index=False)

In [3]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd

data2 = []
for page_number in range(1, 89):
    url = f'https://www.index.hr/oglasi/motorni-brodovi/gid/3573?page={page_number}'

    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title'))
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        price_EUR_tag = ad.find('span', class_='price')
        if price_EUR_tag:
            price_EUR = price_EUR_tag.find('span').text.strip()
        else:
            price_EUR = 'Not stated'
        print(f'Price EUR: {price_EUR}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_link = ad.find('a', class_='result')
        if ad_link:
            ad_url = ad_link['href']
        else:
            ad_url = 'Not stated'
        print(f'Ad URL: {ad_url}\n')

        data2.append({
            'Title': title,
            'Price_EUR': price_EUR,
            'County': county,
            'Ad_URL': ad_url
        })

df = pd.DataFrame(data2)
display(df)

Title: Ocqueteau 615
Price EUR: 14.000 €
County: Splitsko-dalmatinska
Ad URL: https://www.index.hr/oglasi/ocqueteau-615/oid/3934788

Title: Not stated
Price EUR: Not stated
County: Not stated
Ad URL: Not stated

Title: Pasara 6.50
Price EUR: 10.500 €
County: Splitsko-dalmatinska
Ad URL: https://www.index.hr/oglasi/pasara-6-50/oid/3934657

Title: Mariah SC 19
Price EUR: 22.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/mariah-sc-19/oid/3934322

Title: Jeanneau Merry Fisher 795
Price EUR: 77.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/jeanneau-merry-fisher-795/oid/3934289

Title: Quicksilver commander 635
Price EUR: 17.000 €
County: Not stated
Ad URL: https://www.index.hr/oglasi/quicksilver-commander-635/oid/3932897

Title: Not stated
Price EUR: Not stated
County: Not stated
Ad URL: Not stated

Title: Idea Marine 58 Open b+Yamaha 115 KS - ekskl. uvoznik i distributer
Price EUR: 36.035 €
County: Grad Zagreb
Ad URL: https://www.index.hr/oglasi

Title   Price_EUR  \
0                                 Ocqueteau 615    14.000 €   
1                                    Not stated  Not stated   
2                                   Pasara 6.50    10.500 €   
3                                  Mariah SC 19    22.000 €   
4                     Jeanneau Merry Fisher 795    77.000 €   
...                                         ...         ...   
1139  Gliser norveška gradnja + Honda BF 150 4t    13.000 €   
1140            Fisherman 6 metara+Johnson 90ks     6.500 €   
1141                        jeanneau leader 705    23.000 €   
1142                  Bellingardo 630 Sea Ghost     6.500 €   
1143                                 Not stated  Not stated   

                      County  \
0       Splitsko-dalmatinska   
1                 Not stated   
2       Splitsko-dalmatinska   
3         Primorsko-goranska   
4         Primorsko-goranska   
...                      ...   
1139        Šibensko-kninska   
1140  Dubrovačko-neretvanska   
1141      Primorsko-goranska   
1142      Primorsko-goranska   
1143              Not stated   

                                                 Ad_URL  
0     https://www.index.hr/oglasi/ocqueteau-615/oid/...  
1                                            Not stated  
2     https://www.index.hr/oglasi/pasara-6-50/oid/39...  
3     https://www.index.hr/oglasi/mariah-sc-19/oid/3...  
4     https://www.index.hr/oglasi/jeanneau-merry-fis...  
...                                                 ...  
1139  https://www.index.hr/oglasi/gliser-norveska-gr...  
1140  https://www.index.hr/oglasi/fisherman-6-metara...  
1141  https://www.index.hr/oglasi/jeanneau-leader-70...  
1142  https://www.index.hr/oglasi/bellingardo-630-se...  
1143                                         Not stated  

[1144 rows x 4 columns]

In [17]:
df.to_csv('boat_ads2.csv', index=False, encoding='utf-8-sig')

In [18]:
df.to_csv('boat_ads3.csv', sep=',', encoding='utf-8-sig', index=False)

In [22]:
df.to_csv('boat_ads4.csv', sep=',', encoding='utf-8-sig', index=False)
df['Price_EUR'] = df['Price_EUR'].str.replace('€', '')

In [14]:
data2 = []
for page_number in range(1, 89):
    print("processing page" + str(page_number))
    url = f'https://www.index.hr/oglasi/motorni-brodovi/gid/3573?pojam=&sortby=1&elementsNum=10&cijenaod=0&cijenado=2100000&tipoglasa=1&pojamZup=-2&attr_Int_1239=&attr_Int_1240=&vezani_na=1239-1240&num' + str(page_number)
    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title'))
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        price_EUR_tag = ad.find('span', class_='price')
        if price_EUR_tag:
            price_EUR = price_EUR_tag.find('span').text.strip()
        else:
            price_EUR = 'Not stated'
        print(f'Price EUR: {price_EUR}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_link = ad.find('a', class_='result')
        if ad_link:
            ad_url = ad_link['href']
        else:
            ad_url = 'Not stated'
        print(f'Ad URL: {ad_url}\n')

        data2.append({
            'Title': title,
            'Price_EUR': price_EUR,
            'County': county,
            'Ad_URL': ad_url
        })

df = pd.DataFrame(data2)
display(df)

processing page1
Title: ELAN PILOTINA 686
Price EUR: 7.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/elan-pilotina-686/oid/3934881

Title: Not stated
Price EUR: Not stated
County: Not stated
Ad URL: Not stated

Title: Ocqueteau 615
Price EUR: 14.000 €
County: Splitsko-dalmatinska
Ad URL: https://www.index.hr/oglasi/ocqueteau-615/oid/3934788

Title: Pasara 6.50
Price EUR: 10.500 €
County: Splitsko-dalmatinska
Ad URL: https://www.index.hr/oglasi/pasara-6-50/oid/3934657

Title: Mariah SC 19
Price EUR: 22.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/mariah-sc-19/oid/3934322

Title: Jeanneau Merry Fisher 795
Price EUR: 77.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/jeanneau-merry-fisher-795/oid/3934289

Title: Not stated
Price EUR: Not stated
County: Not stated
Ad URL: Not stated

Title: Quicksilver commander 635
Price EUR: 17.000 €
County: Not stated
Ad URL: https://www.index.hr/oglasi/quicksilver-commander-635/oid/393

KeyboardInterrupt: 

In [5]:
df.to_csv('boat_ads6.csv', index=False, encoding='utf-8-sig')

## Working code

In [1]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd
data2 = []
for page_number in range(1, 89):
    url = f'https://www.index.hr/oglasi/motorni-brodovi/gid/3573?pojam=&sortby=1&elementsNum=10&cijenaod=0&cijenado=2100000&tipoglasa=1&pojamZup=-2&attr_Int_1239=&attr_Int_1240=&vezani_na=1239-1240&num={page_number}'
    print(f'Scraping page {page_number}: {url}')
    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title'))
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        price_EUR_tag = ad.find('span', class_='price')
        if price_EUR_tag:
            price_EUR = price_EUR_tag.find('span').text.strip()
        else:
            price_EUR = 'Not stated'
        print(f'Price EUR: {price_EUR}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_link = ad.find('a', class_='result')
        if ad_link:
            ad_url = ad_link['href']
        else:
            ad_url = 'Not stated'
        print(f'Ad URL: {ad_url}\n')

        data2.append({
            'Title': title,
            'Price_EUR': price_EUR,
            'County': county,
            'Ad_URL': ad_url
        })

df = pd.DataFrame(data2)
display(df)

Scraping page 1: https://www.index.hr/oglasi/motorni-brodovi/gid/3573?pojam=&sortby=1&elementsNum=10&cijenaod=0&cijenado=2100000&tipoglasa=1&pojamZup=-2&attr_Int_1239=&attr_Int_1240=&vezani_na=1239-1240&num=1
Title: Fiberstamp - P600
Price EUR: 12.800 €
County: Not stated
Ad URL: https://www.index.hr/oglasi/fiberstamp-p600/oid/3935196

Title: Not stated
Price EUR: Not stated
County: Not stated
Ad URL: Not stated

Title: Adriatic 790
Price EUR: 23.000 €
County: Istarska
Ad URL: https://www.index.hr/oglasi/adriatic-790/oid/3935114

Title: ELAN PILOTINA 686
Price EUR: 7.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/elan-pilotina-686/oid/3934881

Title: Ocqueteau 615
Price EUR: 14.000 €
County: Splitsko-dalmatinska
Ad URL: https://www.index.hr/oglasi/ocqueteau-615/oid/3934788

Title: Pasara 6.50
Price EUR: 10.500 €
County: Splitsko-dalmatinska
Ad URL: https://www.index.hr/oglasi/pasara-6-50/oid/3934657

Title: Not stated
Price EUR: Not stated
County: Not stated
Ad UR

Title   Price_EUR                County  \
0     Fiberstamp - P600    12.800 €            Not stated   
1            Not stated  Not stated            Not stated   
2          Adriatic 790    23.000 €              Istarska   
3     ELAN PILOTINA 686     7.000 €    Primorsko-goranska   
4         Ocqueteau 615    14.000 €  Splitsko-dalmatinska   
...                 ...         ...                   ...   
1139        parasailing    37.162 €  Splitsko-dalmatinska   
1140       Riva replika    50.000 €              Zadarska   
1141   KIIM Sport 23 V6     6.200 €    Primorsko-goranska   
1142  Crownline 210 CCR    18.000 €    Primorsko-goranska   
1143         Not stated  Not stated            Not stated   

                                                 Ad_URL  
0     https://www.index.hr/oglasi/fiberstamp-p600/oi...  
1                                            Not stated  
2     https://www.index.hr/oglasi/adriatic-790/oid/3...  
3     https://www.index.hr/oglasi/elan-pilotina-686/...  
4     https://www.index.hr/oglasi/ocqueteau-615/oid/...  
...                                                 ...  
1139  https://www.index.hr/oglasi/parasailing/oid/63...  
1140  https://www.index.hr/oglasi/riva-replika/oid/3...  
1141  https://www.index.hr/oglasi/kiim-sport-23-v6/o...  
1142  https://www.index.hr/oglasi/crownline-210-ccr/...  
1143                                         Not stated  

[1144 rows x 4 columns]

In [14]:
df['Price_EUR'] = df['Price_EUR'].str.replace('€', '€')
df.to_csv('boat_ads7.csv', index=False)

In [1]:
!pip install psycopg2 sqlalchemy

     ---------------------------------------- 0.0/2.0 MB ? eta -:--:--
     ---- ----------------------------------- 0.2/2.0 MB 6.3 MB/s eta 0:00:01
     ---------- ----------------------------- 0.5/2.0 MB 8.0 MB/s eta 0:00:01
     ------------------- -------------------- 1.0/2.0 MB 8.6 MB/s eta 0:00:01
     ------------------------------- -------- 1.6/2.0 MB 9.9 MB/s eta 0:00:01
     ---------------------------------------  1.9/2.0 MB 8.8 MB/s eta 0:00:01
     ---------------------------------------- 2.0/2.0 MB 8.9 MB/s eta 0:00:00
     ---------------------------------------- 0.0/192.2 kB ? eta -:--:--
     ------------------------------------- 192.2/192.2 kB 11.4 MB/s eta 0:00:00


In [3]:
import psycopg2
from sqlalchemy import create_engine
import pandas as pd

# Define the columns for the boat_ads table
columns = {
    'ID': 'SERIAL PRIMARY KEY',
    'Title': 'VARCHAR(255)',
    'Price_EUR': 'VARCHAR(255)',
    'County': 'VARCHAR(255)',
    'Ad_URL': 'VARCHAR(255)'
}

# Create a connection to the PostgreSQL server
engine = create_engine('postgresql+psycopg2://postgres:1001@localhost:5432/')

# Create a new database
with engine.connect() as conn:
    conn.execute('CREATE DATABASE boat_ads_db')

# Connect to the new database
engine = create_engine('postgresql+psycopg2://postgres:1001@localhost:5432/boat_ads_db')

# Create the boat_ads table
with engine.connect() as conn:
    columns_str = ', '.join([f'{col} {data_type}' for col, data_type in columns.items()])
    conn.execute(f'CREATE TABLE boat_ads ({columns_str})')

# Insert the data from the df DataFrame into the boat_ads table
with engine.connect() as conn:
    df.to_sql('boat_ads', conn, if_exists='append', index=False)


ObjectNotExecutableError: Not an executable object: 'CREATE DATABASE boat_ads_db'

In [4]:
import psycopg2

# Connect to the PostgreSQL server
conn = psycopg2.connect(
    host="localhost",
    port="5432",
    user="postgres",
    password="1001"
)

# Create a new database
with conn:
    with conn.cursor() as cur:
        cur.execute("CREATE DATABASE boat_ads_db")

# Close the connection
conn.close()

ActiveSqlTransaction: CREATE DATABASE cannot run inside a transaction block


In [5]:
import psycopg2
from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
from sqlalchemy import create_engine
import pandas as pd

# Define the columns for the boat_ads table
columns = {
    'ID': 'SERIAL PRIMARY KEY',
    'Title': 'VARCHAR(255)',
    'Price_EUR': 'VARCHAR(255)',
    'County': 'VARCHAR(255)',
    'Ad_URL': 'VARCHAR(255)'
}

# Create a connection to the PostgreSQL server
engine = create_engine('postgresql+psycopg2://postgres:1001@localhost:5432/', isolation_level='AUTOCOMMIT')

# Create a new database
with engine.connect() as conn:
    conn.execute('CREATE DATABASE boat_ads_db')

# Connect to the new database
engine = create_engine('postgresql+psycopg2://postgres:1001@localhost:5432/boat_ads_db')

# Create the boat_ads table
with engine.connect() as conn:
    columns_str = ', '.join([f'{col} {data_type}' for col, data_type in columns.items()])
    conn.execute(f'CREATE TABLE boat_ads ({columns_str})')

# Insert the data from the df DataFrame into the boat_ads table
with engine.connect() as conn:
    df.to_sql('boat_ads', conn, if_exists='append', index=False)

ObjectNotExecutableError: Not an executable object: 'CREATE DATABASE boat_ads_db'

In [6]:
import psycopg2
from sqlalchemy import create_engine, text
import pandas as pd

# Define the columns for the boat_ads table
columns = {
    'ID': 'SERIAL PRIMARY KEY',
    'Title': 'VARCHAR(255)',
    'Price_EUR': 'VARCHAR(255)',
    'County': 'VARCHAR(255)',
    'Ad_URL': 'VARCHAR(255)'
}

# Create a connection to the PostgreSQL server
engine = create_engine('postgresql+psycopg2://postgres:1001@localhost:5432/')

# Create a new database
with engine.connect() as conn:
    conn.execution_options(isolation_level="AUTOCOMMIT").execute(text('CREATE DATABASE boat_ads_db'))

# Connect to the new database
engine = create_engine('postgresql+psycopg2://postgres:1001@localhost:5432/boat_ads_db')

# Create the boat_ads table
with engine.connect() as conn:
    columns_str = ', '.join([f'{col} {data_type}' for col, data_type in columns.items()])
    conn.execute(f'CREATE TABLE boat_ads ({columns_str})')

# Insert the data from the df DataFrame into the boat_ads table
with engine.connect() as conn:
    df.to_sql('boat_ads', conn, if_exists='append', index=False)

ObjectNotExecutableError: Not an executable object: 'CREATE TABLE boat_ads (ID SERIAL PRIMARY KEY, Title VARCHAR(255), Price_EUR VARCHAR(255), County VARCHAR(255), Ad_URL VARCHAR(255))'

In [3]:
import psycopg2
import pandas as pd
!pip install sqlalchemy
from sqlalchemy import create_engine


[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


  Using cached SQLAlchemy-2.0.7-cp310-cp310-win_amd64.whl (2.0 MB)
  Using cached typing_extensions-4.5.0-py3-none-any.whl (27 kB)
  Using cached greenlet-2.0.2-cp310-cp310-win_amd64.whl (192 kB)


In [7]:

df.to_sql('boat_ads', engine, if_exists='append', index=False)

NameError: name 'data2' is not defined

In [1]:
import requests as r
from bs4 import BeautifulSoup as soup
import pandas as pd
data2 = []
for page_number in range(1, 89):
    url = f'https://www.index.hr/oglasi/motorni-brodovi/gid/3573?pojam=&sortby=1&elementsNum=10&cijenaod=0&cijenado=2100000&tipoglasa=1&pojamZup=-2&attr_Int_1239=&attr_Int_1240=&vezani_na=1239-1240&num={page_number}'
    print(f'Scraping page {page_number}: {url}')
    data = r.get(url)
    page_soup = soup(data.text, 'html.parser')
    ads = page_soup.find_all('div', class_='OglasiRezHolder')
    for ad in ads:
        title_tag = ad.find('span', class_=lambda c: c.startswith('title'))
        if title_tag:
            title = title_tag.text.strip()
        else:
            title = 'Not stated'
        print(f'Title: {title}')
        price_EUR_tag = ad.find('span', class_='price')
        if price_EUR_tag:
            price_EUR = price_EUR_tag.find('span').text.strip()
        else:
            price_EUR = 'Not stated'
        print(f'Price EUR: {price_EUR}')
        county_tag = ad.find('li', class_='icon-marker')
        if county_tag:
            county = county_tag.text.strip()
        else:
            county = 'Not stated'
        print(f'County: {county}')
        ad_link = ad.find('a', class_='result')
        if ad_link:
            ad_url = ad_link['href']
        else:
            ad_url = 'Not stated'
        print(f'Ad URL: {ad_url}\n')

        data2.append({
            'Title': title,
            'Price_EUR': price_EUR,
            'County': county,
            'Ad_URL': ad_url
        })

import psycopg2
import pandas as pd
from sqlalchemy import create_engine
df = pd.DataFrame(data2)
display(df)
username = 'postgres'
password = '1001'
host = '172.17.0.2'
port = '5432'
database = 'ads'



# create the connection string
conn_string = 'postgresql://postgres:1001@172.17.0.2:5432/ads'


# create the engine object
engine = create_engine(conn_string)
df.to_sql('boat_ads', engine, if_exists='append', index=False)

Scraping page 1: https://www.index.hr/oglasi/motorni-brodovi/gid/3573?pojam=&sortby=1&elementsNum=10&cijenaod=0&cijenado=2100000&tipoglasa=1&pojamZup=-2&attr_Int_1239=&attr_Int_1240=&vezani_na=1239-1240&num=1
Title: Gliser Rajden 6.90
Price EUR: 0,00 €
County: Dubrovačko-neretvanska
Ad URL: https://www.index.hr/oglasi/gliser-rajden-6-90/oid/3936517

Title: Not stated
Price EUR: Not stated
County: Not stated
Ad URL: Not stated

Title: Prodajem plastičnu barku SEB
Price EUR: 929 €
County: Istarska
Ad URL: https://www.index.hr/oglasi/prodajem-plasticnu-barku-seb/oid/3558876

Title: Fiberstamp - P600
Price EUR: 12.800 €
County: Not stated
Ad URL: https://www.index.hr/oglasi/fiberstamp-p600/oid/3935196

Title: Adriatic 790
Price EUR: 23.000 €
County: Istarska
Ad URL: https://www.index.hr/oglasi/adriatic-790/oid/3935114

Title: ELAN PILOTINA 686
Price EUR: 7.000 €
County: Primorsko-goranska
Ad URL: https://www.index.hr/oglasi/elan-pilotina-686/oid/3934881

Title: Not stated
Price EUR: Not st

Title   Price_EUR                  County  \
0               Gliser Rajden 6.90      0,00 €  Dubrovačko-neretvanska   
1                       Not stated  Not stated              Not stated   
2     Prodajem plastičnu barku SEB       929 €                Istarska   
3                Fiberstamp - P600    12.800 €              Not stated   
4                     Adriatic 790    23.000 €                Istarska   
...                            ...         ...                     ...   
1139                   parasailing    37.162 €    Splitsko-dalmatinska   
1140                  Riva replika    50.000 €                Zadarska   
1141              KIIM Sport 23 V6     6.200 €      Primorsko-goranska   
1142             Crownline 210 CCR    18.000 €      Primorsko-goranska   
1143                    Not stated  Not stated              Not stated   

                                                 Ad_URL  
0     https://www.index.hr/oglasi/gliser-rajden-6-90...  
1                                            Not stated  
2     https://www.index.hr/oglasi/prodajem-plasticnu...  
3     https://www.index.hr/oglasi/fiberstamp-p600/oi...  
4     https://www.index.hr/oglasi/adriatic-790/oid/3...  
...                                                 ...  
1139  https://www.index.hr/oglasi/parasailing/oid/63...  
1140  https://www.index.hr/oglasi/riva-replika/oid/3...  
1141  https://www.index.hr/oglasi/kiim-sport-23-v6/o...  
1142  https://www.index.hr/oglasi/crownline-210-ccr/...  
1143                                         Not stated  

[1144 rows x 4 columns]

OperationalError: (psycopg2.OperationalError) connection to server at "172.17.0.2", port 5432 failed: Connection timed out (0x0000274C/10060)
	Is the server running on that host and accepting TCP/IP connections?

(Background on this error at: https://sqlalche.me/e/20/e3q8)

## Working connection

In [11]:
import psycopg2
import pandas as pd

# Establish a connection to the PostgreSQL database
conn = psycopg2.connect(
    host="localhost",
    port=5432,
    database="ads",
    user="postgres",
    password="1001"
)

# Open the CSV file using Pandas
df4 = pd.read_csv("boat_ads7.csv")

# Insert the data from the CSV file into the database table
cur = conn.cursor()
for index, row in df4.iterrows():
    cur.execute("""
        INSERT INTO boat_ads (title, price_eur, county, ad_url)
        VALUES (%s, %s, %s, %s);
    """, (row['Title'], row['Price_EUR'], row['County'], row['Ad_URL']))

# Commit the changes to the database and close the connection
conn.commit()
conn.close()

In [4]:
import psycopg2
# Connect to the PostgreSQL database
conn = psycopg2.connect(
    host="localhost",
    port="5432",
    database="ads",
    user="postgres",
    password="1001"
)
# Create a cursor object to execute SQL queries
cur = conn.cursor()
# Execute a sample SQL query
cur.execute("SELECT * FROM boat_ads")
# Fetch the query result
result = cur.fetchall()
# Print the query result
print(result)
# Close the cursor and connection objects
cur.close()
conn.close()

[]


In [3]:
import psycopg2
# Connect to the PostgreSQL database
conn = psycopg2.connect(
    host="localhost",
    port="5432",
    database="sibin",
    user="sibinuser",
    password="sib123"
)
# Create a cursor object to execute SQL queries
cur = conn.cursor()
# Execute a sample SQL query
cur.execute("SELECT * FROM users")
# Fetch the query result
result = cur.fetchall()
# Print the query result
print(result)
# Close the cursor and connection objects
cur.close()
conn.close()

[(1, 'John', 'Doe'), (2, 'Jane', 'Doe'), (3, 'Bob', 'Smith')]


In [4]:
import pandas as pd

# load the CSV data into a DataFrame
df8 = pd.read_csv('boat_ads7.csv')

# count the number of rows containing only one or two "not stated" columns
one_not_stated = ((df8 == 'Not stated').sum(axis=1) == 1).sum()
two_not_stated = ((df8 == 'Not stated').sum(axis=1) == 2).sum()
all_not_stated = (df8 == 'Not stated').all(axis=1).sum()


# display the results
print(f'Number of rows containing only one "Not stated" column: {one_not_stated}')
print(f'Number of rows containing only two "Not stated" columns: {two_not_stated}')
print(f'Number of rows where all columns have the value "Not stated": {all_not_stated}')

Number of rows containing only one "Not stated" column: 115
Number of rows containing only two "Not stated" columns: 0
Number of rows where all columns have the value "Not stated": 264


In [10]:

df9 = pd.read_csv('boat_ads7.csv')

# Drop the rows where all columns say "not stated"
df9 = df9[~(df9 == 'Not stated').all(axis=1)]

# Drop the rows where only one or two columns say "not stated"
df9 = df9[((df9 == 'Not stated').sum(axis=1) == 0) | ((df9 == 'Not stated').sum(axis=1) >= 3)]
df9 = df9[df9['Price_EUR'] != '0,00 €']
df9.to_csv('boat_ads9.csv', index=False, encoding='utf-8-sig')

In these cells below I was trying to decide which version of the clean data do I want to load into my database (explained in more detail in the last cell).

In [13]:
import psycopg2
import pandas as pd

# Connect to the PostgreSQL database
conn4 = psycopg2.connect(
    host="localhost",
    port="5432",
    database="ads",
    user="postgres",
    password="1001"
)

# Define the columns for the table
columns = {
    'ID': 'SERIAL PRIMARY KEY',
    'Title': 'VARCHAR(255)',
    'Price_EUR': 'VARCHAR(255)',
    'County': 'VARCHAR(255)',
    'Ad_URL': 'VARCHAR(255)'
}

# Define the SQL statement for creating the table
create_table2 = "CREATE TABLE boat_ads_clean ("

for col_name, col_type in columns.items():
    create_table2 += f"{col_name} {col_type}, "

create_table2 = create_table2[:-2] + ");"

# Create the table in the database
cur2 = conn4.cursor()
cur2.execute(create_table2)

# Open the cleaned CSV file using Pandas
df11 = pd.read_csv("boat_ads9.csv")

# Insert the data from the cleaned DataFrame into the database table
cur2 = conn4.cursor()
for index, row in df11.iterrows():
    cur2.execute("""
        INSERT INTO boat_ads_clean (title, price_eur, county, ad_url)
        VALUES (%s, %s, %s, %s);
    """, (row['Title'], row['Price_EUR'], row['County'], row['Ad_URL']))

# Commit the changes to the database and close the connection
conn4.commit()
conn4.close()

In [17]:
df12 = pd.read_csv('boat_ads9.csv')
df12 = df12[df12['Price_EUR'] > '1.000 €']
df12.to_csv('boat_ads10.csv', index=False)

In [33]:
df12 = pd.read_csv('boat_ads10.csv')
df12 = df12[df12['Price_EUR'] > '1.000 €']
df12.to_csv('boat_ads10.csv', index=False)
# Connect to the PostgreSQL database
conn5 = psycopg2.connect(
    host="localhost",
    port="5432",
    database="ads",
    user="postgres",
    password="1001"
)

# Define the columns for the table
columns = {
    'ID': 'SERIAL PRIMARY KEY',
    'Title': 'VARCHAR(255)',
    'Price_EUR': 'VARCHAR(255)',
    'County': 'VARCHAR(255)',
    'Ad_URL': 'VARCHAR(255)'
}

# Define the SQL statement for creating the table
create_table3 = "CREATE TABLE boat_ads_clean ("

for col_name, col_type in columns.items():
    create_table3 += f"{col_name} {col_type}, "

create_table3 = create_table3[:-2] + ");"

# Create the table in the database
cur3 = conn5.cursor()
cur3.execute(create_table3)

# Open the cleaned CSV file using Pandas
df12 = pd.read_csv("boat_ads10.csv")

# Insert the data from the cleaned DataFrame into the database table
cur3 = conn5.cursor()
for index, row in df11.iterrows():
    cur3.execute("""
        INSERT INTO boat_ads_clean (title, price_eur, county, ad_url)
        VALUES (%s, %s, %s, %s);
    """, (row['Title'], row['Price_EUR'], row['County'], row['Ad_URL']))

# Commit the changes to the database and close the connection
conn5.commit()
conn5.close()

In [34]:
display(df12)

Title Price_EUR  \
0                                         Adriatic 790  23.000 €   
1                                    ELAN PILOTINA 686   7.000 €   
2                                        Ocqueteau 615  14.000 €   
3                                          Pasara 6.50  10.500 €   
4                                         Mariah SC 19  22.000 €   
..                                                 ...       ...   
722  ADRIA 550 PG----GLISER SA KABINOM--VERZIJA ZA ...   4.500 €   
723                            BAYLINER 2755 SUNBRIDGE  22.000 €   
724                                        parasailing  37.162 €   
725                                       Riva replika  50.000 €   
726                                   KIIM Sport 23 V6   6.200 €   

                   County                                             Ad_URL  
0                Istarska  https://www.index.hr/oglasi/adriatic-790/oid/3...  
1      Primorsko-goranska  https://www.index.hr/oglasi/elan-pilotina-686/...  
2    Splitsko-dalmatinska  https://www.index.hr/oglasi/ocqueteau-615/oid/...  
3    Splitsko-dalmatinska  https://www.index.hr/oglasi/pasara-6-50/oid/39...  
4      Primorsko-goranska  https://www.index.hr/oglasi/mariah-sc-19/oid/3...  
..                    ...                                                ...  
722  Splitsko-dalmatinska  https://www.index.hr/oglasi/adria-550-pg-glise...  
723    Primorsko-goranska  https://www.index.hr/oglasi/bayliner-2755-sunb...  
724  Splitsko-dalmatinska  https://www.index.hr/oglasi/parasailing/oid/63...  
725              Zadarska  https://www.index.hr/oglasi/riva-replika/oid/3...  
726    Primorsko-goranska  https://www.index.hr/oglasi/kiim-sport-23-v6/o...  

[727 rows x 4 columns]

In [40]:
df13 = pd.read_csv('boat_ads_11.csv')

# Connect to the PostgreSQL database
conn6 = psycopg2.connect(
    host="localhost",
    port="5432",
    database="ads",
    user="postgres",
    password="1001"
)

# Define the columns for the table
columns = {
    'ID': 'SERIAL PRIMARY KEY',
    'Title': 'VARCHAR(255)',
    'Price_EUR': 'VARCHAR(255)',
    'County': 'VARCHAR(255)',
    'Ad_URL': 'VARCHAR(255)'
}

# Define the SQL statement for creating the table
create_table4 = "CREATE TABLE boat_ads_clean ("

for col_name, col_type in columns.items():
    create_table4 += f"{col_name} {col_type}, "

create_table4 = create_table4[:-2] + ");"

# Create the table in the database
cur4 = conn6.cursor()
cur4.execute(create_table3)

# Open the cleaned CSV file using Pandas
df13 = pd.read_csv("boat_ads_11.csv")

# Insert the data from the cleaned DataFrame into the database table
cur4 = conn6.cursor()
for index, row in df11.iterrows():
    cur4.execute("""
        INSERT INTO boat_ads_clean (title, price_eur, county, ad_url)
        VALUES (%s, %s, %s, %s);
    """, (row['Title'], row['Price_EUR'], row['County'], row['Ad_URL']))

# Commit the changes to the database and close the connection
conn6.commit()
conn6.close()

In [10]:
import pandas as pd
df13 = pd.read_csv('boat_ads10.csv')
display(df13)

Title Price_EUR  \
0                                         Adriatic 790  23.000 €   
1                                    ELAN PILOTINA 686   7.000 €   
2                                        Ocqueteau 615  14.000 €   
3                                          Pasara 6.50  10.500 €   
4                                         Mariah SC 19  22.000 €   
..                                                 ...       ...   
722  ADRIA 550 PG----GLISER SA KABINOM--VERZIJA ZA ...   4.500 €   
723                            BAYLINER 2755 SUNBRIDGE  22.000 €   
724                                        parasailing  37.162 €   
725                                       Riva replika  50.000 €   
726                                   KIIM Sport 23 V6   6.200 €   

                   County                                             Ad_URL  
0                Istarska  https://www.index.hr/oglasi/adriatic-790/oid/3...  
1      Primorsko-goranska  https://www.index.hr/oglasi/elan-pilotina-686/...  
2    Splitsko-dalmatinska  https://www.index.hr/oglasi/ocqueteau-615/oid/...  
3    Splitsko-dalmatinska  https://www.index.hr/oglasi/pasara-6-50/oid/39...  
4      Primorsko-goranska  https://www.index.hr/oglasi/mariah-sc-19/oid/3...  
..                    ...                                                ...  
722  Splitsko-dalmatinska  https://www.index.hr/oglasi/adria-550-pg-glise...  
723    Primorsko-goranska  https://www.index.hr/oglasi/bayliner-2755-sunb...  
724  Splitsko-dalmatinska  https://www.index.hr/oglasi/parasailing/oid/63...  
725              Zadarska  https://www.index.hr/oglasi/riva-replika/oid/3...  
726    Primorsko-goranska  https://www.index.hr/oglasi/kiim-sport-23-v6/o...  

[727 rows x 4 columns]

In [11]:
df14 = pd.read_csv('boat_ads_clean.csv')
display(df14)

id                                              title price_eur  \
0      1                                       Adriatic 790  23.000 €   
1      2                                  ELAN PILOTINA 686   7.000 €   
2      3                                      Ocqueteau 615  14.000 €   
3      4                                        Pasara 6.50  10.500 €   
4      5                                       Mariah SC 19  22.000 €   
..   ...                                                ...       ...   
744  745  ADRIA 550 PG----GLISER SA KABINOM--VERZIJA ZA ...   4.500 €   
745  746                            BAYLINER 2755 SUNBRIDGE  22.000 €   
746  747                                        parasailing  37.162 €   
747  748                                       Riva replika  50.000 €   
748  749                                   KIIM Sport 23 V6   6.200 €   

                   county                                             ad_url  
0                Istarska  https://www.index.hr/oglasi/adriatic-790/oid/3...  
1      Primorsko-goranska  https://www.index.hr/oglasi/elan-pilotina-686/...  
2    Splitsko-dalmatinska  https://www.index.hr/oglasi/ocqueteau-615/oid/...  
3    Splitsko-dalmatinska  https://www.index.hr/oglasi/pasara-6-50/oid/39...  
4      Primorsko-goranska  https://www.index.hr/oglasi/mariah-sc-19/oid/3...  
..                    ...                                                ...  
744  Splitsko-dalmatinska  https://www.index.hr/oglasi/adria-550-pg-glise...  
745    Primorsko-goranska  https://www.index.hr/oglasi/bayliner-2755-sunb...  
746  Splitsko-dalmatinska  https://www.index.hr/oglasi/parasailing/oid/63...  
747              Zadarska  https://www.index.hr/oglasi/riva-replika/oid/3...  
748    Primorsko-goranska  https://www.index.hr/oglasi/kiim-sport-23-v6/o...  

[749 rows x 5 columns]

Between these 727 and 749-row data frames I inserted the 749-row one into my PostgreSQL database. Even though some of those records could be considered outliers (unrealistically low prices, boat parts, or something that is not a boat), there
are still actual boats being sold for under 1000 euros. I dropped or masked all of those "boats" at one point in the second task (for which Mr.Jerković said that it was ok but that there were not enough descriptive variables, so I did my analysis on another dataset) because this solution seemed better for that context. 

The difference between the raw and clean data, in this case, is that the raw data (which I also stored in the PostgreSQL database) has 1144 rows and almost 400 of them have all the values as "not stated" - those are sponsored advertisements. Oh, and I am not ashamed to admit that I worded those as "not stated" because I had originally thought that my code broke 
because it could not process an advertisement in which the seller didn't post an (integral) element, but I
soon realized what the whole deal was about and kept the wording. The code gives it away, but I solved the
problem by using the if/else statements. 